pkg/destinations/destinations.go

```diff
@@ -83,24 +86,71 @@ func (m *DestinationManager) TestCredentials(creds config.Destination) error {
 	return nil
 }
 
-func (m *DestinationManager) Destination(ctx context.Context, databaseID int64) (Destination, error) {
+func (m *DestinationManager) UpdateDestination(ctx context.Context, dest dmodels.Destination, creds map[string]any) error {
+	err := m.storage.Database.UpdateDestination(ctx, dest)
+	if err != nil {
+		return err
+	}
+
+	ds, err := json.Marshal(creds)
+	if err != nil {
+		return err
+	}
+	err = m.storage.Vault.SetCredential(strconv.Itoa(int(dest.ID)), string(ds))
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (m *DestinationManager) CreateDestination(ctx context.Context, teamID uint, dest config.Destination) (uint, error) {
+	// TODO breadchris create a destination in the database and store the credentials in the vault
+	d, err := m.storage.Database.CreateDestination(ctx, teamID, dest.Name, dest.Type)
+	if err != nil {
+		return 0, err
+	}
+
+	ds, err := json.Marshal(dest.Settings)
+	if err != nil {
+		return 0, err
+	}
+	err = m.storage.Vault.SetCredential(strconv.Itoa(int(d.ID)), string(ds))
+	if err != nil {
+		return 0, err
+	}
+	return d.ID, nil
+}
 
+func (m *DestinationManager) Destination(ctx context.Context, databaseID uint) (Destination, error) {
 	if m.mux.TryLock(databaseID) {
 		defer m.mux.Unlock(databaseID)
 
-		var dest Destination
+		var (
+			dest     Destination
+			settings map[string]any
+		)
 
 		dest, ok := m.pool[databaseID]
 		if ok {
 			return dest, nil
 		}
 
+		// XXX breadchris does the caller verify existence and authz for this database?
+		// is this call needed?
 		creds, err := m.storage.Database.GetDestinationCredentials(ctx, databaseID)
```

Comment:
I'd suggest deleting this from the `database` interface completely. If you fix the resulting compiler errors then everything will probably Just Work.

Reply:
This change will be a bit involved since there are multiple callers of `DestinationManager.Destination` where they do not have the database `type` in scope. 

Perhaps a follow up to this one?

Reply:
Now that I've looked in more detail I see how it works. Probably this function should just be renamed `GetDestination`. And yeah, I think we do want to keep this function to make sure the destination actually exists and is valid before trying to fetch credentials from a vault.